Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -4142,9 +4144,149 @@ - (void)setShouldObserveDBModifications:(BOOL)shouldObserveDBModifications
             CGPoint newContentOffset = CGPointMake(0, MAX(0, newContentHeight - viewTopToContentBottom));
             [self.collectionView setContentOffset:newContentOffset animated:NO];
         }
+
+        // When we resume observing database changes, we want to scroll to show the user
+        // any new items inserted while we were not observing.  We therefore find the
+        // first item at or after the "view horizon".  See the comments below which explain
+        // the "view horizon".
+        ConversationViewItem *_Nullable lastViewItem = self.viewItems.lastObject;
+        BOOL hasAddedNewItems = (lastViewItem &&
+            previousLastTimestamp &&
+                                 lastViewItem.interaction.timestamp > previousLastTimestamp.unsignedLongLongValue);
+
+        DDLogInfo(@"%@ hasAddedNewItems: %d", self.logTag, hasAddedNewItems);
+        if (hasAddedNewItems) {
+            NSIndexPath *_Nullable indexPathToShow = [self firstIndexPathAtViewHorizonTimestamp];
+            if (indexPathToShow) {
+                [self.collectionView scrollToItemAtIndexPath:indexPathToShow
+                                            atScrollPosition:UICollectionViewScrollPositionCenteredVertically
```

Comment:
For the sake of longer messages, do we want UICollectionViewScrollPositionTop?

TIOLI

Reply:
Right, it's a tough call, because my goal is to show _both_ the last item before the "view horizon" and the first item after the "view horizon".  We can't do "top on first item after" or "bottom on last item before" or we won't see the other. Unfortunately, this gets tricky if either is huge.  The largest cells are oversize text, which should be rare.  Other cells are considerably smaller than a screenful.  So I think this is reasonable.

Left a comment.